### PR TITLE
More strict commit policy for stable branch

### PR DESCRIPTION
**Pull request by aszlig**
The motivation behind this was a cherry-pick of c204036 to the 15.09 release branch, which broke evaluation. Now I've cherry-picked another commit which was needed by c204036 (58e9440) because otherwise evaluation would fail.

I by myself didn't state a reason at the time where I was cherry-picking the commit but figured that I should have done so after being [noted by @lethalman](https://github.com/NixOS/nixpkgs/commit/fb1377a14aeb06c982098c8b22987708f931743b#commitcomment-14906480) about a related work that removes the changes in the commit I have cherry-picked.

Long story short: If we cherry-pick stuff to any of the release branches, we should explicitly state the reasons _why_ we did so.

Not only because reviewers spend less time wondering about the introduced changes but to give the committer time to rethink about whether it really _is_ necessary to include the change to a supposedly
stable branch.

### CODE REVIEW

**Comment by aszlig** [comment]
Cc: @NixOS/nixpkgs-committers

**Comment by devhell** [comment]
:+1:

**Comment by ttuegel** [comment]
:+1:

File: doc/submitting-changes.xml [hunk 1]
```diff
@@ -276,6 +276,24 @@ Additional information.
...
+</para>
+<para>An example of a cherry-picked commit would look like this:</para>
+<screen>
+nixos: Refactor the world.
+
+The original commit message describing the reason why the world was torn apart.
+
+Cherry-picked-from: abcdef
+Cherry-pick-reason: I just had a gut feeling that this would also be wanted by
+                    people from the stone age.
```
Comment (edolstra) [inline doc/submitting-changes.xml]: We don't need `Cherry-picked-from` because it's already provided by `git cherry-pick -x` (though I see `submitting-changes.xml` doesn't mention `-x`, so we should probably add a note that `-x` is required). So it would be something like:
```
(cherry picked from commit 54d6f1f683a448e094f31b5732c424c2467eaf8d)
Reason: bla bla bla
```
Also, providing a reason is redundant if the original message has an issue link (e.g. `Fixes #123`).

Comment (aszlig) [inline doc/submitting-changes.xml]: Well, I was using `Cherry-picked-from` because it looks more consistent.

If the original message has an issue link it doesn't necessarily mean that it's a justification for cherry-picking it to stable.

**Comment by wizeman** [comment]
As long as we're discussing this, shouldn't there be an actual policy written in that document for what could or should be cherry-picked to the stable branch vs what shouldn't?

**Comment by aszlig** [comment]
@wizeman: That's actually a good point. I mean security and bug fixes may sound obvious, but even that isn't so clear if you give it a second thought, for example: Do you just add a patch or update to the newest upstream version? What about mass-rebuild changes? And what about changes that are not security/bug fixes? What about changes to NixOS modules?

**Comment by DamienCassou** [comment]
Maybe we could agree to adhere to http://chris.beams.io/posts/git-commit/

**Comment by jagajaga** [comment]
@DamienCassou this article is nice, but it's over plus for us IMO.

File: doc/submitting-changes.xml [hunk 2]
```diff
@@ -262,21 +262,69 @@ Additional information.
...
-</listitem>
-</itemizedlist>
+<section>
+  <title>Stable release branches</title>
+
+  <itemizedlist>
+    <listitem>
+      <para>
+        If you're cherry-picking and/or commit to a stable release branch always
+        use <command>git cherry-pick -x</command> add a clear description about
```
Comment (mcmtroffaes) [inline doc/submitting-changes.xml]: The documentation could also suggest to use the ``-e`` flag so the commit message can be edited, i.e.

```xml
<command>git cherry-pick -x -e</command>
```

**Comment by viric** [comment]
I think that the best start is to by example.

Find agreement with a list of *good commits to stable*, and a list of *bad
commits to stable*. I think that any attempt to write a generalization
without examples will not be clear enough.

Please use actual commits that happened; there is no need to write against
what never happened.

Regards,
Lluís.

File: doc/submitting-changes.xml (hunk 2 above)
Comment (vcunat) [inline doc/submitting-changes.xml]: Also, the sentence is relatively hard to read for me. Let me suggest
```xml
If you're cherry-picking a commit to a stable release branch,
always use <command>git cherry-pick -xe</command>
and ensure the message contains a clear description about
why this needs to be included in the stable branch.
```

Comment (mcmtroffaes) [inline doc/submitting-changes.xml]: :+1:

**Comment by aszlig** [comment]
@DamienCassou: I already am quite "nazi" about commit messages and I'm already committing almost as in the linked article. However I do put a period at the end of the subject (will avoid that from now on because I didn't realize that I always OCDed it in).

So from my point of view, I'd strongly recommend making that the default commit policy so that we at least can try to encourage people to do so.

**Comment by garbas** [comment]
:+1:

**Comment by vcunat** [comment]
Squashed and pushed (while leaving authorship with Aszlig).
